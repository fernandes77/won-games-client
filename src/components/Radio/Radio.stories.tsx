import type { Meta, StoryObj } from '@storybook/react'
import { Radio } from './Radio'

const meta: Meta<typeof Radio> = {
  component: Radio
}

export default meta
type Story = StoryObj<typeof Radio>

export const Primary: Story = {
  parameters: {
    layout: 'fullscreen',
    backgrounds: {
      default: 'won-dark'
    }
  },
  decorators: [
    () => (
      <>
        <div style={{ padding: 10 }}>
          <Radio
            label="primeiro"
            labelFor="primeiro"
            id="primeiro"
            name="nome"
            value="primeiro"
            defaultChecked
          />
        </div>
        <div style={{ padding: 10 }}>
          <Radio
            label="segundo"
            labelFor="segundo"
            id="segundo"
            name="nome"
            value="segundo"
          />
        </div>
        <div style={{ padding: 10 }}>
          <Radio
            label="terceiro"
            labelFor="terceiro"
            id="terceiro"
            name="nome"
            value="terceiro"
          />
        </div>
      </>
    )
  ]
}
