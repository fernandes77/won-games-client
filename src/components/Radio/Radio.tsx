'use client'

type RadioValue = string | ReadonlyArray<string> | number

type RadioProps = {
  onCheck?: (value?: RadioValue) => void
  label?: string
  labelColor?: 'white' | 'black'
  labelFor?: string
  value?: RadioValue
} & React.InputHTMLAttributes<HTMLInputElement>

export const Radio = ({
  label,
  onCheck,
  labelColor = 'white',
  labelFor = '',
  value,
  ...props
}: RadioProps) => {
  const onChange = () => {
    !!onCheck && onCheck(value)
  }

  const labelColorClass = { white: 'text-white', black: 'text-black' }

  return (
    <div className="flex items-center">
      <input
        id={labelFor}
        type="radio"
        value={value}
        onChange={onChange}
        className="flex items-center justify-center appearance-none w-[1.125rem] h-[1.125rem]
          border-2 border-solid border-primary rounded-full bg-transparent outline-none
          transition-[background_0.1s_ease-in-out] cursor-pointer
          focus:shadow-[0_0_0.5rem] focus:shadow-primary
          before:w-2 before:h-2 before:rounded-full before:bg-primary
          before:transition-[opacity_0.1s_ease-in-out] before:opacity-0 before:absolute
          checked:before:opacity-100"
        {...props}
      />
      {!!label && (
        <label
          htmlFor={labelFor}
          className={`${labelColorClass[labelColor]} pl-xxs leading-none cursor-pointer`}
        >
          {label}
        </label>
      )}
    </div>
  )
}
