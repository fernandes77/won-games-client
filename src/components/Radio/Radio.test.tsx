import { render, screen, user, waitFor } from '@/utils/tests/helpers'
import { Radio } from './Radio'

describe('<Radio />', () => {
  it('should render with label (white)', () => {
    const { container } = render(
      <Radio label="Radio" labelFor="check" value="anyValue" />
    )

    const label = screen.getByText('Radio')
    expect(label).toBeInTheDocument()
    expect(label).toHaveClass('text-white')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render with label (black)', () => {
    render(<Radio label="Radio" labelColor="black" />)

    const label = screen.getByText('Radio')
    expect(label).toBeInTheDocument()
    expect(label).toHaveClass('text-black')
  })

  it('should render without label', () => {
    render(<Radio />)

    expect(screen.queryByLabelText('Radio')).not.toBeInTheDocument()
  })

  it('should dispatch onCheck when label status changes', async () => {
    const onCheck = jest.fn()
    render(
      <Radio
        label="Radio"
        labelFor="Radio"
        onCheck={onCheck}
        value="anyValue"
      />
    )

    expect(onCheck).not.toHaveBeenCalled()

    user.click(screen.getByLabelText('Radio'))
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1)
    })
    expect(onCheck).toHaveBeenCalledWith('anyValue')
  })

  it('Should be accessible with tab', async () => {
    render(<Radio label="Radio" labelFor="Radio" />)

    const radio = screen.getByLabelText('Radio')

    expect(document.body).toHaveFocus()

    await user.tab()

    expect(radio).toHaveFocus()
  })
})
