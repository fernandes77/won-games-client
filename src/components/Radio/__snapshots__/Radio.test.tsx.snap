// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> should render with label (white) 1`] = `
<div
  class="flex items-center"
>
  <input
    class="flex items-center justify-center appearance-none w-[1.125rem] h-[1.125rem] border-2 border-solid border-primary rounded-full bg-transparent outline-none transition-[background_0.1s_ease-in-out] cursor-pointer focus:shadow-[0_0_0.5rem] focus:shadow-primary before:w-2 before:h-2 before:rounded-full before:bg-primary before:transition-[opacity_0.1s_ease-in-out] before:opacity-0 before:absolute checked:before:opacity-100"
    id="check"
    type="radio"
    value="anyValue"
  />
  <label
    class="text-white pl-xxs leading-none cursor-pointer"
    for="check"
  >
    Radio
  </label>
</div>
`;
